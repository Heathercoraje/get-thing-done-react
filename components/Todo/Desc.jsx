import React, { Component } from 'react';
import ReactDOM from 'react';
import styled from 'styled-components';

const DescBox = styled.div`
	margin-bottom: 3vh;
`;

class Desc extends Component {
	render() {
		const name = this.props.name;
		const description = this.props.description;
		return (
			<DescBox className="desc-box">
				<DescForm
    name={name}
					description={description}
    HandleDesc={this.props.HandleDesc}
  />
  </DescBox>
		);
	}
}

class DescForm extends Component {
	state = {
		edit: false
	};
	componentWillReceiveProps() {
		this.setState({
			edit: false
		});
	}

	toggle = () => {
		const prev = this.state.edit;
		this.setState({
			edit: !prev
		});
	};

	render() {
		const name = this.props.name;
		const description = this.props.description;
		if (this.state.edit) {
			return (
				<div>
					<Header name={name} toggle={this.toggle} />
					<form
						className="form-desc"
						onSubmit={evt => {
							this.props.HandleDesc(evt);
							this.toggle();
						}}
					>
						<input
							size="22"
							className="input-desc"
							type="text"
							placeholder={description}
							onBlur={evt => this.props.HandleDesc(evt)}
							autoFocus
						/>
						<input style={{ display: 'none' }} type="submit" />
					</form>
				</div>
			);
		} else {
			return (
				<div>
					<Header name={name} toggle={this.toggle} />
					<div id="desc" onClick={this.toggle}>
						{description}
					</div>
				</div>
			);
		}
	}
}

const Header = props => {
	return (
		<div className="cate-title">
			<h1 className="todo-name">{props.name}</h1>
			<button type="button" className="button-edit" onClick={props.toggle}>
				&nbsp;&#10000;
			</button>
		</div>
	);
};

export { Desc, DescForm, Header };
