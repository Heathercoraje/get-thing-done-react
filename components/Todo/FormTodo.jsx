import React from 'react';
import styled from 'styled-components';
import HandleClickInput from '../../lib/HandleClickInput';

const Wrapper = styled.div`
	margin-top: auto;
`;

const FormTodo = ({ add }) => {
	const onClickInput = evt => {
		const inputTodo = document.getElementById('input-todo');
		evt.preventDefault();
		HandleClickInput(evt, add, inputTodo);
	};
	return (
		<Wrapper>
			<form onSubmit={onClickInput}>
				<input
					className="input-gen input-todo"
					id="input-todo"
					style={{ display: 'none' }}
					placeholder="Enter a new task"
				/>
				<input type="submit" value="+" className="button-add" />
			</form>
		</Wrapper>
	);
};

export default FormTodo;
