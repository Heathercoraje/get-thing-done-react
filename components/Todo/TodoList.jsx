import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { Desc } from './Desc';

const Wrapper = styled.div`
	width: 100%;
	overflow-y: auto;
	overflow-x: hidden;
`;

const TodoList = props => {
	const todos = props.todos;
	const selected = props.selected;
	const name = props.name;
	const descText = props.categories[selected].description;
	const Renderlist = !todos
		? ''
		: todos.map(todo => (
				<TodoItem
					todo={todo.todo}
					key={todo.id}
					id={todo.id}
					isDone={todo.isDone}
					value={todo.id}
					delete={props.delete}
					complete={props.complete}
				/>
			));
	return (
		<Wrapper>
			<Desc name={name} description={descText} HandleDesc={props.HandleDesc} />
			<div>{Renderlist}</div>
		</Wrapper>
	);
};

TodoList.propTypes = {
	todos: PropTypes.array,
	categories: PropTypes.array,
	selected: PropTypes.number,
	name: PropTypes.string,
	delete: PropTypes.func,
	complete: PropTypes.func,
	HandleDesc: PropTypes.func
}
const TodoItem = props => (
	<div key={props.id} className="list-todo">
		<button className="checkbox" onClick={props.delete} value={props.value}>
			&#9744;
		</button>
		<li
			id={props.id}
			onClick={props.complete}
			value={props.value}
			style={!props.isDone ?
			null : { textDecoration: 'line-through' }}
		>
			{props.todo}
		</li>
	</div>
);

TodoItem.propTypes = {
	todo: PropTypes.string,
	id: PropTypes.number,
	value: PropTypes.string,
	isDone: PropTypes.bool,
	delete: PropTypes.func,
	complete: PropTypes.func
}

export { TodoList, TodoItem };
