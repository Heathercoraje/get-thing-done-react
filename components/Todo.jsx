import React, { Component } from 'react';
import styled from 'styled-components';
import TodoList from './TodoList';
import FormTodo from './FormTodo';

const Wrapper = styled.div`
	box-sizing: border-box;
	border-top-right-radius: 0.2rem;
	border-bottom-right-radius: 0.2rem;
	background: white;
	padding: 3rem 1.5rem 1.5rem 2.5rem;
	float: right;
	width: 65%;
	height: 100%;
	display: flex;
	flex-direction: column;
	color: #393939;
`;

const HandleDesc = (e) => {
  if (e.keyCode === 13) {
    document.getElementById('descForm').submit();
  }
};

class Todo extends Component {
  constructor(props) {
    super(props);
    this.state = {
      description: 'hello world'
    };
    this.HandleDesc = this.HandleDesc.bind(this);
  }
  HandleDesc(e) {
    if (e.keyCode === 13) {
      this.setState({
        description: e.target.value
      });
      e.preventDefault();
    }
  }
  render() {
    return (
      <Wrapper>
        <TodoList
          categories={this.props.categories}
          todos={this.props.todos}
          selected={this.props.selected}
          delete={this.props.delete}
          HandleDesc={this.HandleDesc}
          desc={this.state.description}
        />
        <FormTodo add={this.props.add} selected={this.props.selected} />
      </Wrapper>
    );
  }
}

export default Todo;
