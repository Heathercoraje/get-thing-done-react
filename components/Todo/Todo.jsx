import React, { Component } from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { TodoList } from './TodoList';
import FormTodo from './FormTodo';
import localStorageApi from '../../lib/localStorageApi';


const Wrapper = styled.div`
  padding: 3rem 1.5rem 1.5rem 2.5rem;
	background: white;
	display: flex;
	flex-direction: column;
	color: #393939;
`;
class Todo extends Component {
	state = {
		todos: {}
	};

  componentWillMount () {
    localStorageApi.loadTodos().then((todos)=>{
      console.log('Fetching your data, friend.');
      this.setState({todos});
    });
  }

	componentWillReceiveProps(Nextprops) {
		this.updateTodoObj(Nextprops);
	}

	updateTodoObj = Nextprops => {
		const todoObj = Object.assign({}, this.state.todos);
		const updatedCategories = Nextprops.categories.map(
			category => category.name
		);
		for (const name in todoObj) {
			if (updatedCategories.indexOf(name) < 0) {
				delete todoObj[name];
			}
			continue;
		}
    localStorageApi.modifyTodos(todoObj);
		this.setState({
			todos: todoObj
		});
	};
	addTodo = evt => {
		const allTodos = this.state.todos;
		const selected = this.props.selected;
		const category = this.props.categories[selected];
		const name = category.name;
		// if this is a new categoty add new key of category and add empty array
		if (!allTodos[`${name}`]) {
			allTodos[`${name}`] = [];
		}
		const Todo = allTodos[`${name}`];
		const newTodo = {
			id: Todo.length === 0 ? 0 : Todo[Todo.length - 1].id + 1,
			todo: evt.target[0].value,
			isDone: false
		};
		Todo.push(newTodo);
    localStorageApi.modifyTodos(allTodos).then(()=>{
      this.setState({
        todos: allTodos
      });
      console.log('todo item added');
    });
	};
	deleteTodo = evt => {
		const allTodos = this.state.todos;
		const selected = this.props.selected;
		const category = this.props.categories[selected];
		const name = category.name;
		const newTodos = allTodos[`${name}`].filter(
			todo => todo.id !== Number(evt.target.value)
		);
		allTodos[`${name}`] = newTodos;
    localStorageApi.modifyTodos(allTodos).then(()=>{
      this.setState({
        todos: allTodos
      });
      console.log('todo item deleted');
    });

	};
	completeTodo = evt => {
		const allTodos = this.state.todos;
		const selected = this.props.selected;
		const category = this.props.categories[selected];
		const name = category.name;
		const targetTodoArr = allTodos[`${name}`];
		const targetId = evt.target.id;
		targetTodoArr.forEach(todo => {
			if (todo.id == targetId) {
				let temp = todo.isDone;
				todo.isDone = !temp;
			}
		});
		allTodos[`${name}`] = targetTodoArr;
    localStorageApi.modifyTodos(allTodos).then(()=>{
      this.setState({
        todos: allTodos
      });
      console.log('todo item completed');
    });

	};

	render() {
		const selected = this.props.selected;
		if (selected < 0) {
			return (
				<Wrapper>
					<h3> Add a new category </h3>{' '}
				</Wrapper>
			);
		}
		const category = this.props.categories[selected];
		const name = category.name;
		return (
			<Wrapper className="todo-container">
				<TodoList
					todos={this.state.todos[`${name}`]}
					selected={selected}
					name={name}
					categories={this.props.categories}
					HandleDesc={this.props.HandleDesc}
					delete={this.deleteTodo}
					complete={this.completeTodo}
    />{' '}
				<FormTodo add={this.addTodo} />{' '}
			</Wrapper>
		);
	}
}

Todo.propTypes = {
  categories: PropTypes.array,
  selected: PropTypes.number,
  HandleDesc: PropTypes.func,
}

export default Todo;
