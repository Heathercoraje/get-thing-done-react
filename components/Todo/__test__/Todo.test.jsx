import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import Todo from '../Todo';


test('Todo renders correctly', () => {
  const props = {
    categories :[{id:0, name:'shopping', description:'shopping list'}],
    selected : 0
  }
	const component = shallow(<Todo {...props} />);
	const tree = toJson(component);
  expect(component.find('TodoList').length).toBe(1);
  expect(component.find('FormTodo').length).toBe(1);
  expect(tree).toMatchSnapshot();
});


test('when categorylist is empty, it returns `Add new category` message', () => {
  const props = {
    categories :[],
    selected : -1
  }
	const component = shallow(<Todo {...props} />);
	const tree = toJson(component);
	expect(tree).toMatchSnapshot();
});
