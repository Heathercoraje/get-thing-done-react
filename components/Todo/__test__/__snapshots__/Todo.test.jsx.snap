// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Todo renders correctly 1`] = `
<styled.div
  className="todo-container"
>
  <TodoList
    categories={
      Array [
        Object {
          "description": "shopping list",
          "id": 0,
          "name": "shopping",
        },
      ]
    }
    complete={[Function]}
    delete={[Function]}
    name="shopping"
    selected={0}
  />
   
  <FormTodo
    add={[Function]}
    selected={0}
  />
   
</styled.div>
`;

exports[`when categorylist is empty, it returns \`Add new category\` message 1`] = `
<styled.div>
  <h3>
     Add a new category 
  </h3>
   
</styled.div>
`;
